### DISCUSSION TITLE
ASP.NET Core Roadmap for .NET 10

### DISCUSSION THREAD

**Discussion by mkArtakMSFT** [dotnet/aspnetcore#59444] [General]
Discussion thread on the [ASP.NET Core Roadmap for .NET 10](https://github.com/dotnet/aspnetcore/issues/59443)

**Comment by ghost** [comment]
I would like to see some progress around Observability.

Native instrumentation support:
- https://github.com/dotnet/aspnetcore/issues/52439

Remove some log scopes that no longer make sense with OpenTelemetry and span attributes:
- https://github.com/dotnet/aspnetcore/issues/31028
- https://github.com/dotnet/aspnetcore/pull/44873

Ability to add attributes to the root activity to facilitate Head Sampling:
- https://github.com/dotnet/aspnetcore/issues/50488

**Comment by boukenka** [comment]
I would like to see some progress for the [Consider shipping ObjectGraphDataAnnotationsValidator / ValidateComplexTypeAttribute](https://github.com/dotnet/aspnetcore/issues/28640) issue, which has been open since _December 14, 2020_ (**53** up votes so far)

**Comment by hez2010** [comment]
I would like to see blazor being fully trimming compatible in .NET 10: https://github.com/dotnet/aspnetcore/issues/49409
As it's the prerequisites of building blazor client apps with NativeAOT-LLVM.

We have experimented NativeAOT-LLVM with Avalonia wasm apps and the performance is way faster than mono-aot, also the binary size is also smaller than mono-aot by a lot. Here is an example of a NativeAOT-LLVM compiled Avalonia app: [sizoscopeX](https://hez2010.github.io/sizoscopeX/), you can test this app with [sample-data.zip](https://github.com/user-attachments/files/18115897/sample-data.zip) by selecting both the two files in the archive in the Open dialog to see how fast it is.

I expect blazor to have even better experience, as blazor doesn't use external libraries (like libSkia) for rendering so that the binary size should be a lot smaller than Avalonia.

**Reply by BrunoBlanes** [reply]
I'd also like to see MVC get support for trimming #41767, at least for controllers, since minimal APIs isn't a fit for everyone.

**Reply by KennethHoff** [reply]
> since minimal APIs isn't a fit for everyone.

How so? Other than upgrading legacy applications - which presumably wouldn't use Native AOT anyways, and even if you did, migrating to Minimal API would be the least of your concerns - what is keeping you off of Minimal API?

**Reply by mikekistler** [reply]
@BrunoBlanes Can you say more about why minimal APIs "isn't a fit for everyone"? This is not to disregard your comment, but to understand "the problem underneath the problem".

**Comment by CrahunGit** [comment]
I would like to see Signalr from blazor server working with authentication (cookies or jwt). It can share the signalr connection used by render updates.

**Comment by rogihee** [comment]
My personal wishlist:
1. https://github.com/dotnet/aspnetcore/issues/52216 (it was on the Roadmap for 8 and 9 but not yet on 10): 
2. https://github.com/dotnet/aspnetcore/issues/42561 (this hurts a lot when toggling between pre-rendering and rendering)
3. https://github.com/dotnet/aspnetcore/issues/44586 (makes library initializers not work)
4. Razor compiler should fail if a component has an attribute that is not a parameter and component is "pure" (no AdditionalAttributes parameter catch-all defined). This causes the most runtime regressions / errors that are hard to detect when migrating between versions of external component libraries. 
5. Routing that supports multiple languages (ideally: a source generator that generates unique routes that can be referenced in code as well, independent of language, with support for resx-based slugs)

**Comment by HectorOrtizTorresInnomotics** [comment]
I would like to have more functionality in QuickGrid for example the ability to implement conditional formatting (by rows or by columns) without losing PropertyColumn filtering and sorting functionality.  TemplateColumn kind of allows to do the conditional formatting but you lose the Filtering and Sorting Functionalities.  Can we have all the PropertyColumn goods in the TemplateColumn or the ability to have ChildContent in the PropertyColumn.

**Comment by boukenka** [comment]
I would like to see some progress for the [(BlazorServer) each navigation to external logs 2 errors a minute later](https://github.com/dotnet/aspnetcore/issues/45267) issue. Which is closed but should be reopened. And also the possibility to catch exceptions from the circuit.

**Comment by coder-johnz** [comment]
I would like to see improvements of Blazor Webassembly
1. Real multithreading: https://github.com/dotnet/aspnetcore/issues/17730
2. Improve startup performance: https://github.com/dotnet/aspnetcore/issues/42284

**Comment by ndavidb** [comment]
I would like to see Blazor Server automatically preserve the components' state and event handlers when there is a session state loss in the app. According to Dan's appearances on YouTube, it sounds like you are already considering this a main focus area.

**Comment by Kumima** [comment]
- https://github.com/dotnet/aspnetcore/issues/35501
- https://github.com/dotnet/aspnetcore/issues/53189

Improvement for minimal api.

**Comment by ahmed-abdelrazek** [comment]
I would like to see microsoft identity server supporting an option for decrypted jwt access token for the front end to read the claims
and an option to required user to have unique phone number
![image](https://github.com/user-attachments/assets/5c9f442d-db3c-48b6-89c1-d2cbf5d533fb)

**Reply by mikekistler** [reply]
@ahmed-abdelrazek These are some interesting ideas. Are there already issues for these? If not could you file them?

**Reply by ahmed-abdelrazek** [reply]
Require Unique Phone Number https://github.com/dotnet/aspnetcore/issues/55199
Unprotected JWT Access Token for Bearer Scheme https://github.com/dotnet/aspnetcore/issues/60667

@mikekistler

**Comment by Tsholofelo-Nkuna** [comment]
I would like to have routable blazor components make use of the "route" or introduce an "@route" attribute to configure their routes  as opposed to making use of the "@page" attribute; 

If a blazor component makes use of a "@page" attribute, I would expect it  to have all if not most capabilities that a razor page has or even more.

**Reply by danroth27** [reply]
Hi @Tsholofelo-Nkuna. Is this to differentiate pages in a Blazor app from pages in a Razor Pages? Note that Blazor components already use the .razor extension instead of the .cshtml extension used for MVC & Razor Pages.

**Reply by Tsholofelo-Nkuna** [reply]
Yes, correct routable  blazor components make use of the .razor extension as well as the @page (to specify the route)when declaring them ; however they do not have all functionality that a normal page would have, i.e. direct access **(I'm aware that you can inject IHttpContextAccessor  to get the HttpContext)** to HttpContext. so I would rather not declare them using @page as they are not pages but are components; If I want a component to be routable, I thought it would make more sense just to introduce an @route to specify the route instead of using @page(to specify the route).

**Reply by danroth27** [reply]
Since we added static server-side rendering (SSR) to Blazor in .NET 8, you should be able to do pretty much anything from a routable Blazor component that you could do from a .cshtml based Razor Page. For example, you can get access to the HttpContext from a Blazor component as a cascading parameter assuming the component is rendering in the context of a request, like this:

```csharp
[CascadingParameter]
public HttpContext? HttpContext { get; set; }
```

We reused `@page` in Blazor for routable components because the semantics are similar: the file represents a page that you can navigate to from a browser. Since switching from `@page` to `@route` would be a breaking change, I don't think that's something we would do at this point. I also don't think we'd introduce `@route` in addition to `@page` because having multiple was to do the same thing often just results in confusion.

But, if there are other things that you think you should be able to do from a page in Blazor that you currently can't do, we could consider adding support for those scenarios.

**Comment by mguinness** [comment]
Can [DisplayNameFor support in Blazor](https://github.com/dotnet/aspnetcore/issues/49147) be considered for .NET 10?  If others agree I'd recommend upvoting that issue to show interest.

**Comment by valterlei-viana** [comment]
I would like to see some improvements about Blazor WASM without API tier, but accessing DB like MSSQL Server.

**Reply by danroth27** [reply]
@valterlei-viana Can you open a GitHub issue with some more details about this scenario and how you'd like the framework to handle it?

**Comment by valterlei-viana** [comment]
I would like to see a template for Blazor WASM with SQL Lite inside the browser like this project
https://github.com/JeremyLikness/SqliteWasmHelper

**Reply by danroth27** [reply]
Hi @valterlei-viana. We have a pretty high bar for adding new project templates because of the maintenance cost associated with them. Is there a reason why docs and a sample isn't sufficient for this scenario? Note that the .NET template engine is also extensible, so if you wanted to [create a project template](https://learn.microsoft.com/en-us/dotnet/core/tutorials/cli-templates-create-project-template) for this and make it available as a NuGet package for the community that's certainly possible.

**Comment by lakani** [comment]
I would like to see this 

https://github.com/dotnet/runtime/issues/94351

**Reply by danroth27** [reply]
Thanks for the feedback, @lakani! We use that runtime issue to track the progress of post MVP WebAssembly specs and our position on them. Where appropriate, we will leverage these specs as they become available in browsers. If there are specific specs that you are interested in, I recommend commenting and giving a :+1: on the corresponding issue.

**Comment by LuohuaRain** [comment]
I'd like to see this

**[Provide a pure code model for creating and working with Blazor components]**
https://github.com/dotnet/aspnetcore/issues/44866

**Reply by danroth27** [reply]
Thank for the feedback, @LuohuaRain. Unfortunately, for .NET 10 this isn't an area we plan to invest in, but please be sure the note your interest by giving a :+1: reaction on the issues original post.

**Comment by brianmed** [comment]
I'm very grateful for the current HTTP/3 via QUIC implementation used by WebTransport. Is it technically possible to take HTTP/3 and WebTransport out of beta? If so, will there be time?

**Comment by WeihanLi** [comment]
Hope we could configure forwarded behavior from the configuration without updating the code, hope for some feedback

https://github.com/dotnet/aspnetcore/issues/57017

**Comment by vgallegob** [comment]
I would like to see improvements in blazor SSR, event handling like hydro

**Comment by Atulin** [comment]
Allow me to repost a question from [ASP.NET Core Roadmap for .NET 9](https://github.com/dotnet/aspnetcore/discussions/53665#discussioncomment-10861173) thread:

> I've been reading about the new static assets handling (`.MapStaticAssets()`) a little, and I wonder if I will be able to replace `.UseStaticFiles()` with it. I wasn't able to find the following information:
> 1. Can it be configured to handle non-default file types? For example, I currently have a mapping for `.avif` files to `image/avif` MIME type. I could not find any docs about whether it will be necessary and/or possible to do.
> 2. Custom compression algorithms. I currently use an implementation of ZStandard compression, followed by Brotli and GZip, should the browser not support zstd. From what I can tell, the build-time compression will only produce the latter two. Will I be able to add zstd support? Or, if not, is it planned?

**Comment by mguinness** [comment]
A better venue to get an answer to your questions is asking on [Stack Overflow](https://stackoverflow.com/questions/tagged/asp.net-core).  However, have you tried setting `StaticFileOptions` before calling `MapStaticAssets`?

```c#
builder.Services.Configure<StaticFileOptions>(options =>
{
    var provider = new FileExtensionContentTypeProvider();
    provider.Mappings[".avif"] = "image/avif";
    options.ContentTypeProvider = provider;
});
```
